test/hummingbot/connector/exchange/ascend_ex/test_ascend_ex_api_order_book_data_source.py

```diff
@@ -0,0 +1,582 @@
+                    "0.0681",
+                    "43641"
+                ],
+                "bid": [
+                    "0.0676",
+                    "443"
+                ]
+            }]
+        }
+
+        trading_pairs = self.ev_loop.run_until_complete(self.do_fetch_trading_pairs(mock_response))
+
+        self.assertEqual(3, len(trading_pairs))
+        self.assertEqual("BTC-USDT", trading_pairs[1])
+
+    @aioresponses()
+    async def do_get_last_traded_prices_requests_rest_api_price_when_subscription_price_not_available(self, mock_response, api_mock):
+        self.data_source._shared_client = aiohttp.ClientSession()
+        self.data_source._trading_pairs = ["BTC-USDT"]
+
+        url = re.escape(f"{CONSTANTS.REST_URL}/{CONSTANTS.TRADES_PATH_URL}?symbol=")
+        regex_url = re.compile(f"^{url}")
+        api_mock.get(regex_url, body=json.dumps(mock_response))
+
+        results = await self.data_source.get_last_traded_prices(client=self.data_source._shared_client, trading_pairs=[self.trading_pair], throttler=self.throttler)
+
+        await self.data_source._shared_client.close()
+        return results
+
+    def test_get_last_traded_prices_requests_rest_api_price_when_subscription_price_not_available(self):
+        mock_response = {
+            "code": 0,
+            "data": {
+                "m": "trades",
+                "symbol": "BTC/USDT",
+                "data": [
+                    {
+                        "seqnum": 144115191800016553,
+                        "p": "0.06762",
+                        "q": "400",
+                        "ts": 1573165890854,
+                        "bm": False
+                    },
+                    {
+                        "seqnum": 144115191800070421,
+                        "p": "0.06797",
+                        "q": "341",
+                        "ts": 1573166037845,
+                        "bm": True
+                    }
+                ]
+            }
+        }
+
+        results = self.ev_loop.run_until_complete(self.do_get_last_traded_prices_requests_rest_api_price_when_subscription_price_not_available(mock_response))
+        self.assertEqual(results[self.trading_pair], float(mock_response["data"]["data"][1]["p"]))
+
+    @patch('aiohttp.ClientSession.ws_connect', new_callable=AsyncMock)
+    async def do_listen_for_subscriptions_registers_to_orders_trades_and_instruments(self, ws_connect_mock):
+        self.data_source._shared_client = aiohttp.ClientSession()
+        self.data_source._trading_pairs = ["BTC-USDT"]
+        ws_connect_mock.return_value = self.mocking_assistant.create_websocket_mock()
+
+        queue = asyncio.Queue()
+
+        try:
+            self.listening_task = self.ev_loop.create_task(self.data_source.listen_for_trades(ev_loop=self.ev_loop, output=queue))
+            await asyncio.wait_for(self.mocking_assistant._all_ws_delivered.wait(), 0)
+        except asyncio.exceptions.TimeoutError:
+            pass
```

Comment:
The test should not wait for timeouts. It should be created to wait for particular responses or results to be present to continue (for example wait for message responses to appear in the queue).